extensions/azurecore/src/account-provider/utils/msalCachePlugin.ts

```diff
@@ -18,35 +19,54 @@ export class MsalCachePluginProvider {
 		Logger.verbose(`MsalCachePluginProvider: Using cache path ${_msalFilePath} and serviceName ${_serviceName}`);
 	}
 
+	private getLockfilePath(): string {
+		return this._msalFilePath + '.lock';
+	}
+
 	public getCachePlugin(): ICachePlugin {
+		const lockFilePath = this.getLockfilePath();
 		const beforeCacheAccess = async (cacheContext: TokenCacheContext): Promise<void> => {
-			let exists = true;
+			let isAccessible = true;
+			await this.WaitAndLock(lockFilePath);
 			try {
-				await fsPromises.access(this._msalFilePath, constants.R_OK | constants.W_OK);
-			} catch {
-				exists = false;
-			}
-			if (exists) {
 				try {
+					await fsPromises.access(this._msalFilePath, constants.R_OK | constants.W_OK);
+				} catch {
+					Logger.error(`MsalCachePlugin: Cache file is not accessible.`);
+					isAccessible = false;
+				}
+				if (isAccessible) {
 					const cache = await fsPromises.readFile(this._msalFilePath, { encoding: 'utf8' });
 					cacheContext.tokenCache.deserialize(cache);
 					Logger.verbose(`MsalCachePlugin: Token read from cache successfully.`);
-				} catch (e) {
-					Logger.error(`MsalCachePlugin: Failed to read from cache file. ${e}`);
+				}
+			} catch (e) {
+				Logger.error(`MsalCachePlugin: Failed to read from cache file: ${e}`);
+				// Handle JSON syntax error in cache file in case anything goes wrong.
+				if (e.stack.includes('SyntaxError')) {
```

Comment:
This seems risky to assume - what if the phrase SyntaxError shows up somewhere else in the stack? Or the error is changed?

Is there any reason you wouldn't want to just put a try/catch around the deserialize and clear it if that fails for any reason? That would ensure that any issues with parsing the content is handled so is much more future proof too.

Reply:
I can try that